packages/core/src/v3-compat.ts

```diff
@@ -0,0 +1,318 @@
+                }
+
+                if (request.type !== authenticator.type) {
+                    throw new Err(
+                        ErrorCode.AUTHENTICATION_FAILED,
+                        "The auth request type and authenticator type do not match!"
+                    );
+                }
+
+                const provider = this._getAuthServer(request.type);
+
+                try {
+                    await provider.verifyAuthRequest(authenticator, request, { code });
+                    request.status = AuthRequestStatus.Verified;
+                    request.verified = new Date();
+                    authenticator.lastUsed = new Date();
+                    await this.storage.save(auth);
+                } catch (e) {
+                    request.tries++;
+                    await this.storage.save(auth);
+                    throw e;
+                }
+
+                const hasAccount = auth.accountStatus === AccountStatus.Active;
+
+                const hasLegacyAccount = !!this.legacyServer && !!(await this.legacyServer.getStore(email));
+
+                // If the user doesn't have an account but does have a legacy account,
+                // repurpose the verification token for signup
+                if (!hasAccount && hasLegacyAccount) {
+                    request.purpose = AuthPurpose.Signup;
+                }
+
+                let legacyToken: string | undefined = undefined;
+                if (hasLegacyAccount) {
+                    const getLegacyDataMFARequest = new AuthRequest({
+                        type: AuthType.Email,
+                        purpose: AuthPurpose.GetLegacyData,
+                    });
+                    await getLegacyDataMFARequest.init();
+                    getLegacyDataMFARequest.verified = new Date();
+                    getLegacyDataMFARequest.status = AuthRequestStatus.Verified;
+                    legacyToken = getLegacyDataMFARequest.token;
+                    auth.authRequests.push(getLegacyDataMFARequest);
+                }
+
+                await this.storage.save(auth);
+
+                return new RetrieveMFATokenResponse({
+                    token: request.token,
+                    hasAccount,
+                    hasLegacyAccount,
+                    legacyToken,
+                });
+            } catch (e) {
+                throw e;
+            }
+        }
+
+        /**
+         * Initiate the login procedure for a given account by requesting the authentication params
+         * which are required for proceeding with [[createSession]].
+         * @deprecated
+         */
+        @Handler(InitAuthParams, InitAuthResponse)
+        async initAuth({ email, verify }: InitAuthParams): Promise<InitAuthResponse> {
+            const { accountId, srpId, keyParams, B } = await this.startCreateSession(
+                new StartCreateSessionParams({ email, authToken: verify })
+            );
+            srpIds.set(accountId, srpId);
```

Comment:
This is kind of an ugly workaround, but it'll do. The problem is that the v4 api requires the client to hold on to the id of the srp session, which is required by the `completeCreateSession` endpoint. What we're doing here is have the compatibility layer remember the srp id for the client by mapping it to the account id. This has the downside that only one session can be created at a time using this endpoint, but in practice that shouldn't be a problem.